core/jobs/batch_jobs/refresher_exp_id_jobs.py

```diff
@@ -0,0 +1,172 @@
+                    yield (exp.id, state_name)
+                    break
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        """Returns a PCollection of exploration id to state name and user email
+        results from matching refresher_exp_id as not None in exploration.
+
+        Returns:
+            PCollection. A PCollection of 'exp_id: %s, data: %s' results from
+            matching refresher_exp_id as not None in exploration.
+        """
+        exp_to_state_name = (
+            self.pipeline
+            | 'Get all Exploration models' >> ndb_io.GetModels(
+                exp_models.ExplorationModel.get_all())
+            | 'Get exploration from model' >> beam.Map(
+                exp_fetchers.get_exploration_from_model)
+            | 'Extract' >> beam.FlatMap(
+                self._process_exploration_states)
+        )
+
+        exp_rights_collection = (
+            self.pipeline
+            | 'Get all exp rights model' >> ndb_io.GetModels(
+                exp_models.ExplorationRightsModel.get_all())
+            | 'Add exp id as key' >> beam.WithKeys( # pylint: disable=no-value-for-parameter
+                lambda exp_rights: exp_rights.id
+            )
+        )
+
+        relevant_exp_ids = (
+            exp_to_state_name
+            | 'extract just the ids' >> beam.Keys() # pylint: disable=no-value-for-parameter
+        )
+
+        relevant_exp_ids_iter = beam.pvalue.AsIter(relevant_exp_ids)
+
+        exp_rights_filtered = (
+            exp_rights_collection
+            | 'Filter exp rights that have refresher ID' >> beam.Filter(
+                lambda exp_rights, relevant_exp_ids: (
+                    exp_rights[0] in relevant_exp_ids),
+                    relevant_exp_ids=relevant_exp_ids_iter
+            )
+        )
+
+        user_id_to_exp_id = (
+            exp_rights_filtered
+            | beam.FlatMapTuple(self._flatten_user_id_to_exp_id)
+        )
+
+        user_id_to_user_emails = (
+            self.pipeline
+            | 'Get all user settings models' >> ndb_io.GetModels(
+                user_models.UserSettingsModel.get_all())
+            | 'Extract id and email' >> beam.Map(
+                    lambda user_setting: (
+                        user_setting.id, user_setting.email))
+        )
+
+        grouped_exp_id_to_email = (
+            (user_id_to_exp_id, user_id_to_user_emails)
+            | 'Group by user id' >> beam.CoGroupByKey()
+            | 'Drop user id' >> beam.Values() # pylint: disable=no-value-for-parameter
+            | 'Filter out empty results' >> beam.Filter(
+                lambda groups: len(groups[0]) > 0)
+            # Apache beam tries to unpack a, b so it is passed as a tuple.
+            | 'Form tuples' >> beam.FlatMapTuple(
+                lambda a, b: ((a, b),))
+            | 'Flatten exp id list' >> beam.FlatMapTuple(
+                self._flatten_exp_id_to_email)
+            )
```

Comment:
I don't understand the operations here, can it be simplified?

Reply:
I have worked on it but couldn't get them as this is too complicated for me to work with as I am new to Beam jobs. Sorry!!
Is there anything which I can do so that this PR would be merged or will it be reassigned to someone else ??

Reply:
@EricZLou I think this is one example of the `CoGroupByKey` issue.

Reply:
@gopivaibhav I think you can submit testing request anyway. This comment is not that important.